python/python/bystro/proteomics/annotation_interface.py

```diff
@@ -0,0 +1,195 @@
+    keep_alive: str = ONE_DAY
+
+
+def _get_num_slices(
+    client: OpenSearch,
+    index_name: str,
+    opensearch_query_options: OpenSearchQueryOptions,
+    query: dict[str, Any],
+) -> int:
+    """Count number of hits for the index."""
+    get_num_slices_query = query.copy()
+    get_num_slices_query.pop("sort", None)
+    get_num_slices_query.pop("track_total_hits", None)
+
+    response = client.count(body=get_num_slices_query, index=index_name)
+
+    n_docs: int = response["count"]
+    if n_docs < 1:
+        err_msg = (
+            f"Expected at least one document in `response['count']`, got response: {response} instead."
+        )
+        raise RuntimeError(err_msg)
+
+    num_slices_necessary = math.ceil(n_docs / opensearch_query_options.max_query_size)
+    num_slices_planned = min(num_slices_necessary, opensearch_query_options.max_slices)
+    return max(num_slices_planned, 1)
+
+
+def _run_annotation_query(
+    job_data: SaveJobData,
+    client: OpenSearch,
+    opensearch_query_options: OpenSearchQueryOptions,
+) -> pd.DataFrame:
+    """Given query and index contained in SaveJobData, run query and return  in dataframe."""
+
+    query = job_data.queryBody
+    num_slices = _get_num_slices(client, job_data.indexName, opensearch_query_options, query)
+    point_in_time = client.create_point_in_time(  # type: ignore[attr-defined]
+        index=job_data.indexName, params={"keep_alive": opensearch_query_options.keep_alive}
+    )
+    try:  # make sure we clean up the PIT index properly no matter what happens in this block
+        pit_id = point_in_time["pit_id"]
+        query["pit"] = {"id": pit_id}
+        query["size"] = opensearch_query_options.max_query_size
+        remote_queries = []
+        for slice_id in range(num_slices):
+            slice_query = query.copy()
+            if num_slices > 1:
+                # Slice queries require max > 1
+                slice_query["slice"] = {"id": slice_id, "max": num_slices}
+            query_args = {"body": slice_query}
+            remote_query = _process_query(  # type: ignore[call-arg]
+                client,
+                query_args=query_args,
+            )
+            remote_queries.append(remote_query)
+    except Exception as e:
+        err_msg = (
+            f"Encountered exception: {repr(e)} while running opensearch_query, "
+            "deleting PIT index and exiting.\n"
+            f"job_data: {job_data}\n"
+            f"client: {client}\n"
+            f"opensearch_query_options: {opensearch_query_options}\n"
+        )
+        logger.exception(err_msg, exc_info=e)
+        client.delete_point_in_time(body={"pit_id": pit_id})  # type: ignore[attr-defined]
+        raise
+    client.delete_point_in_time(body={"pit_id": pit_id})  # type: ignore[attr-defined]
+    return pd.concat(remote_queries)
+
+
+def _build_opensearch_query_from_query_string(query_string: str) -> dict[str, Any]:
+    return {
+        "query": {
```

Comment:
Ideally we'd be selecting pre-filtering the results to only return the necessary fields, but I keep running into errors like: 

```opensearchpy.exceptions.RequestError: RequestError(400, 'parsing_exception', 'request does not support [_source]')```

see e.g. here for additional context: https://stackoverflow.com/questions/9605292/make-elasticsearch-only-return-certain-fields